### PR TITLE
DirectVelocityPrimitive Should Not Use FW Estimated Robot State

### LINKED ISSUE #1223: DirectVelocityPrimitive should not use firmware estimated robot state
### Description of the task

The implementation of the `DirectVelocity` firmware primitive applies velocity relative to the robot state as tracked by firmware. If this internal state ever drifts, the applied velocities will no longer match the physical state of the robot (ie. applying forward velocity would actually induce sideways velocity). This is particularly noticeable when controlling robots with the xbox controller.

The `DirectVelocity` primitive should apply velocities relative to the physical robot state so that behavior is consistent at all times.

To be clear, this change needs to be made to the primitive implementation in the `firmware` folder.

### Acceptance criteria

- [ ] Change the firmware implementation of the `DirectVelocity` primitive to always apply velocity based on the physical robot state, not the internal state

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

`DirectVelocityPrimitive` should not use the robot state tracked by firmware. It should apply velocities relative to physical robot state to ensure consistent behaviour.

### Testing Done

It builds successfully. Will need to pass "MatCI" later on.

### Resolved Issues

Resolves #1223 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Ah so something I just realized looking lower down in the diff, is that it seems like `trackVelocity` is trying to track _global_ velocity vectors. Since the `trackVelocity` function is only used for the `DirectVelocity` primitive (@kelvinkoon please confirm this is the case), I think we can change it to `trackLocalVelocity`.

Then we can remove the translation from the robot frame to the world frame, since that's the part using the estimated robot state (that we don't want to use). See line 181 in `app/control/control.c`. Then we should be able to do nearly the same calculations before calling `applyAccel`, just using all data in the robot frame.

@kelvinkoon lmk if that makes sense.

**Comment by kelvinkoon** [comment]
> Ah so something I just realized looking lower down in the diff, is that it seems like `trackVelocity` is trying to track _global_ velocity vectors. Since the `trackVelocity` function is only used for the `DirectVelocity` primitive (@kelvinkoon please confirm this is the case), I think we can change it to `trackLocalVelocity`.
> 
> Then we can remove the translation from the robot frame to the world frame, since that's the part using the estimated robot state (that we don't want to use). See line 181 in `app/control/control.c`. Then we should be able to do nearly the same calculations before calling `applyAccel`, just using all data in the robot frame.
> 
> @kelvinkoon lmk if that makes sense.

Can confirm after grepping that `app_control_trackVelocity` is only used in `direct_velocity_primitive`.

I've removed the translation in `control.c`. Going back to beginning of `trackLocalVelocity`, did we want to use the `app_firmware_robot_getVelocityX(robot)` or did we just want to assign `current_vx` to use `linear_velocity_x`?

File: src/firmware/app/control/control.h [hunk 1]
```diff
@@ -25,5 +25,5 @@ void app_control_applyAccel(const FirmwareRobot_t* robot, float linear_accel_x,
  *                          coordinate frame (m/s)
  * @param angular_velocity The angular velocity to move at (rad/s)
  */
-void app_control_trackVelocity(FirmwareRobot_t* robot, float linear_velocity_x,
-                               float linear_velocity_y, float angular_velocity);
+void app_control_trackLocalVelocity(FirmwareRobot_t* robot, float linear_velocity_x,
```
Comment (MathewMacDougall) [inline src/firmware/app/control/control.h]: Javadoc comment should be updated

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just one nit. I will run MatCI :tm: to try this out on a robot

File: src/firmware/app/control/control.h (hunk 1 above)
Comment (kelvinkoon) [reply src/firmware/app/control/control.h]: Done.

Comment (garethellis0) [inline src/firmware/app/control/control.h]: nit; but I would prefer `app_control_trackVelocityInLocalFrame` as a function name.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Just one small nit; but then this looks good to me.

File: src/firmware/app/control/control.h (hunk 1 above)
Comment (kelvinkoon) [reply src/firmware/app/control/control.h]: Done.

**Review by jonathanlew — DISMISSED** [review]
LGTM pending MatCI :tm: and CI

**Comment by garethellis0** [comment]
Hmm, I tried to take a look at the log, but got an error message saying `There was an error trying to fetch the log`. Re-triggered CI, hopefully that should fix it.

**Review by garethellis0 — DISMISSED** [review]
LGTM pending CI :+1: :tada:

File: src/firmware/app/primitives/direct_velocity_primitive.c [hunk 2]
```diff
@@ -32,9 +32,9 @@ static void direct_velocity_tick(void* void_state_ptr, FirmwareWorld_t* world)
     DirectVelocityPrimitiveState_t* state =
         (DirectVelocityPrimitiveState_t*)void_state_ptr;
     FirmwareRobot_t* robot = app_firmware_world_getRobot(world);
-    app_control_trackVelocity(robot, state->direct_target_velocity_x,
-                              state->direct_target_velocity_y,
-                              state->direct_target_velocity_angular);
+    app_control_trackLocalVelocity(robot, state->direct_target_velocity_x,
```
Comment (akhilveeraghanta) [inline src/firmware/app/primitives/direct_velocity_primitive.c]: `app_control_trackLocalVelocity` -> `app_control_trackVelocityInLocalFrame`

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Nice work! Just a minor typo, think that's also the reason why CI is failing

File: src/firmware/app/primitives/direct_velocity_primitive.c (hunk 2 above)
Comment (kelvinkoon) [reply src/firmware/app/primitives/direct_velocity_primitive.c]: Good catch! Fixed.

**Review by akhilveeraghanta — DISMISSED** [review]
LGTM :shipit:

**Comment by MathewMacDougall** [comment]
MatCI :tm: results, using the old mrf-test

Applying angular velocity causes major oscillations and instability. The wheels spin back and forth with increasing magnitude. Even applying a small angular velocity gradually causes larger and larger oscillations that go out of control quickly. See the attached video.

[wheel_oscillation_video.zip](https://github.com/UBC-Thunderbots/Software/files/4471090/wheel_oscillation_video.zip)

X and Y velocity seem ok, but I couldn't do thorough testing due to the issue with angular velocity.

**Comment by MathewMacDougall** [comment]
just realized this may be because `app_firmware_robot_getVelocityX` returns data in the global frame, so we're fixing reference frames in our calculations.

**Comment by kelvinkoon** [comment]
> just realized this may be because `app_firmware_robot_getVelocityX` returns data in the global frame, so we're fixing reference frames in our calculations.

I should simply use `linear_velocity_x` from the input param and assign it to `current_vx`, right? wasn't sure because that'd make `desired_acceleration` be 0.

**Comment by MathewMacDougall** [comment]
I think that would make `desired_acceleration` 0. You should check if `app_firmware_robot_getVelocityX` gives the robot velocity in global or local frame. If it's global, you'll need to translate it to local before doing `desired_velocity` calculations.

**Comment by kelvinkoon** [comment]
> I think that would make `desired_acceleration` 0. You should check if `app_firmware_robot_getVelocityX` gives the robot velocity in global or local frame. If it's global, you'll need to translate it to local before doing `desired_velocity` calculations.

From `firmware_robot.c`, it's in the global frame. Do I use matrix multiplication to translate it? (i.e `current_vx = current_vx * (cosf(current_orientation) * linear_velocity_x)`)

**Comment by MathewMacDougall** [comment]
Take a look at how its done elsewhere in the code (hint: you removed some frame translation code)

**Comment by kelvinkoon** [comment]
@MathewMacDougall kindly requesting some MatCI :tm:

**Comment by MathewMacDougall** [comment]
MatCI results

I'm seeing the same issue as before. I also noticed it happens when applying x and y velocity (which makes sense). It does something similar where it starts jerking back and forth.

I added some code to run the DirectVelocityPrimitive in a simulated test. Check out the latest commit [in this branch](https://github.com/UBC-Thunderbots/Software/commits/pr/1259) (feel free to cherry-pick).

I added a `test_direct_velocity_primitive` in the `simulated_tests_test.cpp` file. You can set the `DirectVelocityPrimitive` in `mock_ai_wrapper.cpp`.

This should let you run / debug stuff more easily and also look at the resulting movement. Hopefully will make this easier to diagnose.

**Comment by kelvinkoon** [comment]
I tested my recent changes using the tests from the other branch to demonstrate the robot rotates without weird behaviour. After people are done checking it using the tests, I'll remove them and push it again for review.

**Comment by MathewMacDougall** [comment]
Checked it and it seems good in simulation on my end. @kelvinkoon did you try other test cases like starting the robot in different positions / orientations, and applying different velocities?

**Comment by kelvinkoon** [comment]
> Checked it and it seems good in simulation on my end. @kelvinkoon did you try other test cases like starting the robot in different positions / orientations, and applying different velocities?

I've tried rotating at different velocities but not trying different starting positions. I'll do those later tonight and report back.

**Comment by kelvinkoon** [comment]
> Checked it and it seems good in simulation on my end. @kelvinkoon did you try other test cases like starting the robot in different positions / orientations, and applying different velocities?

Starting from different positions and orientations is fine, as is applying different velocities.
However, giving it a `y_velocity` seems to make the robot veer left or right when it should be driving straight. The `x_velocity` is fine, however, so I'm not certain why this is happening.

**Comment by MathewMacDougall** [comment]
> However, giving it a y_velocity seems to make the robot veer left or right when it should be driving straight. The x_velocity is fine, however, so I'm not certain why this is happening.

You mean when only applying y_velocity, or y_velocity in combination with something else? If it's not veering sharply, that could just be the behaviour of the controller / simulator since it's just constantly trying to apply a local acceleration, so may not necessarily track in a straight line globally due to the wheel layout and mechanics.

**Comment by kelvinkoon** [comment]
> > However, giving it a y_velocity seems to make the robot veer left or right when it should be driving straight. The x_velocity is fine, however, so I'm not certain why this is happening.
> 
> You mean when only applying y_velocity, or y_velocity in combination with something else? If it's not veering sharply, that could just be the behaviour of the controller / simulator since it's just constantly trying to apply a local acceleration, so may not necessarily track in a straight line globally due to the wheel layout and mechanics.

It was with only applying y_velocity. 
Otherwise, the rest of the testing seemed to be fine, so I'll remove tests and prepare for a 2nd review if it's okay?

**Comment by MathewMacDougall** [comment]
Yea should be good to go :+1:

**Comment by MathewMacDougall** [comment]
Would you like MatCI :tm: or not yet?

**Comment by kelvinkoon** [comment]
> Would you like MatCI :tm: or not yet?

Yes, please!

**Comment by MathewMacDougall** [comment]
MatCI :tm: passed! :heavy_check_mark: 

No funky control issues, and everything worked as expected. There's no more drift in the velocity directions! :heart_eyes:

**Review by MathewMacDougall — DISMISSED** [review]

File: src/firmware/app/control/control.h [hunk 3]
```diff
@@ -16,14 +16,16 @@ void app_control_applyAccel(const FirmwareRobot_t* robot, float linear_accel_x,
...
+ * @param linear_velocity_x The velocity to move at in the x-direction in the robot
  *                          coordinate frame (m/s)
- * @param linear_velocity_y The velocity to move at in the y-direction in the global
+ * @param linear_velocity_y The velocity to move at in the y-direction in the robot
  *                          coordinate frame (m/s)
  * @param angular_velocity The angular velocity to move at (rad/s)
  */
-void app_control_trackVelocity(FirmwareRobot_t* robot, float linear_velocity_x,
-                               float linear_velocity_y, float angular_velocity);
+void app_control_trackVelocityInLocalFrame(FirmwareRobot_t* robot,
```
Comment (akhilveeraghanta) [inline src/firmware/app/control/control.h]: Not a blocking comment, but we seem to say `robot frame` for the robots frame of reference and `global frame` for the frame of reference using the fields coordinate system

For sake of consistency should this be `trackVelocityInRobotFrame`? Not blocking, just a minor comment

**Review by akhilveeraghanta — DISMISSED** [review]

File: src/firmware/app/control/control.h (hunk 3 above)
Comment (jonathanlew) [reply src/firmware/app/control/control.h]: Changing this to trackVelocityInRobotFrame is blocking for me

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nice job! However, Akhil brought up a really good point

**Comment by kelvinkoon** [comment]
> Nice job! However, Akhil brought up a really good point

Changed.

File: src/firmware/app/control/control.h (hunk 3 above)
Comment (kelvinkoon) [reply src/firmware/app/control/control.h]: Done.

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**